### PR TITLE
Learning rate schedules improvement + extension

### PR DESCRIPTION
**Pull request by lukovnikov**
re: [PR#389](https://github.com/huggingface/pytorch-pretrained-BERT/pull/389)
- refactored learning rate schedules into objects
- added `WarmupCosineWithHardRestartsSchedule` for cosine schedule with hard restarts
- added `WarmupCosineWithWarmupRestartsSchedule` for cosine schedule with restarts where each restart uses the same warmup slope

### CODE REVIEW

File: pytorch_pretrained_bert/optimization.py [hunk 1]
```diff
@@ -23,30 +23,145 @@
...
-    return 1.0
-
-def warmup_linear(x, warmup=0.002):
-    """ Specifies a triangular learning rate schedule where peak is reached at `warmup`*`t_total`-th (as provided to BertAdam) training step.
-        After `t_total`-th training step, learning rate is zero. """
-    if x < warmup:
-        return x/warmup
-    return max((x-1.)/(warmup-1.), 0)
+
+__all__ = ["LRSchedule", "WarmupLinearSchedule", "WarmupConstantSchedule", "WarmupCosineSchedule", "BertAdam",
```
Comment (marpaia) [inline pytorch_pretrained_bert/optimization.py]: I believe a declaration of `__all__` should be in the modules `__init__.py`. I also believe that the `pytorch_pretrained_bert` module includes more than just this list which may make this change misleading.

File: pytorch_pretrained_bert/optimization.py [hunk 2]
```diff
@@ -23,30 +23,145 @@
...
+        # end warning
+        return ret
+
+    def get_lr_(self, progress):
+        """
+        :param progress:    value between 0 and 1 (unless going beyond t_total steps) specifying training progress
+        :return:            learning rate multiplier for current update
+        """
+        return 1.
+        # raise NotImplemented("use subclass")  -
```
Comment (marpaia) [inline pytorch_pretrained_bert/optimization.py]: Is this supposed to be commented here? If this is an abstract class, consider using the `abc` module to enforce implementation by subclasses.

File: tests/optimization_test.py [hunk 3]
```diff
@@ -21,6 +21,9 @@
 import torch
 
 from pytorch_pretrained_bert import BertAdam
+from pytorch_pretrained_bert.optimization import WarmupCosineWithWarmupRestartsSchedule
+#from matplotlib import pyplot as plt
```
Comment (marpaia) [inline tests/optimization_test.py]: Is this import not necessary? If so, we should remove it.

File: tests/optimization_test.py [hunk 4]
```diff
@@ -46,5 +49,23 @@ def test_adam(self):
         self.assertListAlmostEqual(w.tolist(), [0.4, 0.2, -0.5], tol=1e-2)
 
 
+class WarmupCosineWithRestartsTest(unittest.TestCase):
+    def test_it(self):
+        m = WarmupCosineWithWarmupRestartsSchedule(warmup=0.05, t_total=1000., cycles=5)
+        x = np.arange(0, 1000)
+        y = [m.get_lr(xe) for xe in x]
+        # plt.plot(y)
+        # plt.show(block=False)
```
Comment (marpaia) [inline tests/optimization_test.py]: If we no longer want to do this rendering in this test, we should remove this code from the test IMO.

File: pytorch_pretrained_bert/optimization.py (hunk 1 above)
Comment (thomwolf) [reply pytorch_pretrained_bert/optimization.py]: I agree. Currently `pytorch-pretrained-bert` is not designed to be used with `import *` (we should update the rest of the library). So it's better to remove this `__all__` to avoid mis-understandings.

**Comment by thomwolf** [comment]
Sorry for the delay in reviewing this.
This is a great PR and it looks good to me.
Thanks for adding some tests also.
I agree with the (mostly cosmetic) comments from @marpaia.
Do you think you can fix them and then we can merge?

**Comment by lukovnikov** [comment]
Addressed @marpaia 's comments.

**Comment by marpaia** [comment]
Awesome @lukovnikov, I think it looks great! Thanks for making this already awesome library even more awesome 💯

**Comment by thomwolf** [comment]
Thanks @lukovnikov!
